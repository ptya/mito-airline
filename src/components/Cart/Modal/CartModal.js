import React, { useContext } from "react";
import PropTypes from "prop-types";

// components
import CartFlight from "../Flight/CartFlight";

// context
import { CartContext } from "components/providers/CartProvider";
import { StationsContext } from "components/providers/StationsProvider";

// utils
import { formatMoney } from "utils/formatMoney";

// global elements
import Modal from "components/_elements/Modal";

// local styles
import Header from "./_styles/Header";
import FlightsWrapper from "./_styles/FlightsWrapper";
import Footer from "./_styles/Footer";
import CancelBtn from "./_styles/CancelBtn";

const CartModal = props => {
  const { setToggle, animation } = props;
  const { cart, cartDispatch } = useContext(CartContext);
  const {
    stations: { origin, destination }
  } = useContext(StationsContext);

  const handleReset = () => {
    setToggle(false);
    cartDispatch({
      type: "purge"
    });
  };

  const { total, inbound, outbound } = cart;

  return (
    <Modal setToggle={setToggle} animation={animation}>
      <Header>Thanks for buying your tickets at Mito Airlines</Header>
      <FlightsWrapper>
        {outbound && (
          <CartFlight flight={outbound} from={origin} to={destination} />
        )}
        {inbound && (
          <CartFlight flight={inbound} from={destination} to={origin} />
        )}
      </FlightsWrapper>
      <Footer>
        <h2>
          TOTAL: <span>{formatMoney(total)}</span>
        </h2>
        <CancelBtn type="button" onClick={() => handleReset()}>
          No, thanks (reset)
        </CancelBtn>
      </Footer>
    </Modal>
  );
};

CartModal.propTypes = {
  setToggle: PropTypes.func.isRequired,
  animation: PropTypes.object.isRequired
};

export default CartModal;
