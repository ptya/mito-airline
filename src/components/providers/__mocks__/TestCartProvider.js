import React, { useReducer } from "react";
import PropTypes from "prop-types";

import { CartContext, cartReducer, initialCart } from "../CartProvider";

const CartProvider = props => {
  const { state, reducer } = props;
  const [cart, cartDispatch] = useReducer(reducer, state);

  return (
    <CartContext.Provider
      value={{
        cart,
        cartDispatch
      }}
    >
      {props.children}
    </CartContext.Provider>
  );
};

CartProvider.propTypes = {
  children: PropTypes.node.isRequired,
  state: PropTypes.object,
  reducer: PropTypes.func
};

CartProvider.defaultProps = {
  state: initialCart,
  reducer: cartReducer
};

export default CartProvider;
