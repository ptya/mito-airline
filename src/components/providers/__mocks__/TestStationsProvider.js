import React, { useReducer } from "react";
import PropTypes from "prop-types";

import {
  StationsContext,
  stationsReducer,
  initialStations
} from "../StationsProvider";

const TestStationsProvider = props => {
  const { state, reducer } = props;
  const [stations, stationsDispatch] = useReducer(reducer, state);

  return (
    <StationsContext.Provider
      value={{
        stations,
        stationsDispatch
      }}
    >
      {props.children}
    </StationsContext.Provider>
  );
};

TestStationsProvider.propTypes = {
  children: PropTypes.node.isRequired,
  state: PropTypes.object,
  reducer: PropTypes.func
};

TestStationsProvider.defaultProps = {
  state: initialStations,
  reducer: stationsReducer
};

export default TestStationsProvider;
