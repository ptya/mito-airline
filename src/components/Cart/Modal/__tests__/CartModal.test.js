import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";

// components
import CartModal from "../CartModal";

// context
import TestCartProvider from "components/providers/__mocks__/TestCartProvider";
import TestStationsProvider from "components/providers/__mocks__/TestStationsProvider";

// mocks
import {
  stationOne,
  stationTwo,
  flight
} from "components/providers/__mocks__/mockValues";

afterEach(() => {
  cleanup();
  console.error.mockClear();
  setToggle.mockClear();
});

console.error = jest.fn();
const setToggle = jest.fn();

function renderModal(cartState) {
  const stationsState = {
    origin: stationOne,
    destination: stationTwo
  };
  const context = render(
    <TestStationsProvider state={stationsState}>
      <TestCartProvider state={cartState}>
        <CartModal setToggle={setToggle} animation={{}} />
      </TestCartProvider>
    </TestStationsProvider>
  );
  return context;
}

test("Renders with 2 stations without errors and matches snapshot", () => {
  renderModal({
    inbound: flight,
    outbound: flight,
    total: 100
  });
  expect(document.body).toMatchSnapshot();
  expect(console.error).not.toHaveBeenCalled();
});

test("Renders with 1 station without errors and matches snapshot", () => {
  renderModal({
    inbound: null,
    outbound: flight,
    total: 100
  });
  expect(document.body).toMatchSnapshot();
  expect(console.error).not.toHaveBeenCalled();
});

test("Clicking on link toggles modal", () => {
  const { getByTestId } = renderModal({
    inbound: flight,
    outbound: flight,
    total: 100
  });
  expect(setToggle).not.toHaveBeenCalled();
  fireEvent.click(getByTestId("cm-btn"));
  expect(setToggle).toHaveBeenCalled();
});
