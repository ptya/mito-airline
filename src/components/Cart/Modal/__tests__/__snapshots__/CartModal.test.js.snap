// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders with 1 station without errors and matches snapshot 1`] = `
<body
  style="overflow: hidden;"
>
  <div />
  <div
    data-testid="my-portal"
  >
    <div
      class="sc-EHOje bDDmvD"
      y="0"
    >
      <main
        class="sc-bZQynM koBpX"
      >
        <h1
          class="sc-gzVnrw dtbiwe"
        >
          Thanks for buying your tickets at Mito Airlines
        </h1>
        <section
          class="sc-htoDjs eDcOzR"
        >
          <div
            class="sc-bdVaJa jmlIJf"
            data-testid="cart-flight"
          >
            <p
              class="sc-bwzfXH cWAdWB"
            >
              <span>
                Jul
              </span>
              <span>
                12
              </span>
            </p>
            <p
              class="sc-htpNat hPUCVW"
            >
              <span>
                Barcelona
              </span>
              <span>
                Budapest
              </span>
              <span>
                Fri
                 
                10:30
                 – 
                13:30
              </span>
            </p>
          </div>
        </section>
        <section
          class="sc-dnqmqq hqaXBR"
        >
          <h2>
            TOTAL: 
            <span>
              $1
            </span>
          </h2>
          <button
            class="sc-iwsKbI lguKDE"
            data-testid="cm-btn"
            type="button"
          >
            No, thanks (reset)
          </button>
        </section>
      </main>
    </div>
  </div>
</body>
`;

exports[`Renders with 2 stations without errors and matches snapshot 1`] = `
<body
  style="overflow: hidden;"
>
  <div />
  <div
    data-testid="my-portal"
  >
    <div
      class="sc-EHOje bDDmvD"
      y="0"
    >
      <main
        class="sc-bZQynM koBpX"
      >
        <h1
          class="sc-gzVnrw dtbiwe"
        >
          Thanks for buying your tickets at Mito Airlines
        </h1>
        <section
          class="sc-htoDjs eDcOzR"
        >
          <div
            class="sc-bdVaJa jmlIJf"
            data-testid="cart-flight"
          >
            <p
              class="sc-bwzfXH cWAdWB"
            >
              <span>
                Jul
              </span>
              <span>
                12
              </span>
            </p>
            <p
              class="sc-htpNat hPUCVW"
            >
              <span>
                Barcelona
              </span>
              <span>
                Budapest
              </span>
              <span>
                Fri
                 
                10:30
                 – 
                13:30
              </span>
            </p>
          </div>
          <div
            class="sc-bdVaJa jmlIJf"
            data-testid="cart-flight"
          >
            <p
              class="sc-bwzfXH cWAdWB"
            >
              <span>
                Jul
              </span>
              <span>
                12
              </span>
            </p>
            <p
              class="sc-htpNat hPUCVW"
            >
              <span>
                Budapest
              </span>
              <span>
                Barcelona
              </span>
              <span>
                Fri
                 
                10:30
                 – 
                13:30
              </span>
            </p>
          </div>
        </section>
        <section
          class="sc-dnqmqq hqaXBR"
        >
          <h2>
            TOTAL: 
            <span>
              $1
            </span>
          </h2>
          <button
            class="sc-iwsKbI lguKDE"
            data-testid="cm-btn"
            type="button"
          >
            No, thanks (reset)
          </button>
        </section>
      </main>
    </div>
  </div>
</body>
`;
